File: androidApp/src/main/java/com/commuto/interfacemobile/android/keymanager/types/PublicKey.kt
package com.commuto.interfacemobile.android.keymanager.types

import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DERNull
import org.bouncycastle.asn1.pkcs.PKCSObjectIdentifiers
import org.bouncycastle.asn1.x509.AlgorithmIdentifier
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.spec.MGF1ParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.security.PublicKey as JavaSecPublicKey
import javax.crypto.Cipher
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PSource

/**
 * This is a wrapper around the [JavaSecPublicKey] class with added support for Commuto Interface
 * IDs and several methods for encoding the wrapped public key. The wrapped key is the public key of
 * a 2048-bit RSA private key, and the interface ID of a public key is the SHA-256 hash of its
 * PKCS#1 formatted byte representation.
 *
 * @property interfaceId The interface ID of this [PublicKey], which is the SHA-256 hash of its
 * PKCS#1 byte representation.
 * @property publicKey the [JavaSecPublicKey] that this class wraps.
 */
//TODO: PublicKey getter
class PublicKey {

    /**
     * Creates a [PublicKey] using the PKCS#1-formatted byte representation of an RSA public key.
     *
     * @param publicKeyBytes The PKCS#1 bytes of the public key to be created, as a [ByteArray].
     */
    constructor(publicKeyBytes: ByteArray) {
        val algorithmIdentifier = AlgorithmIdentifier(
            PKCSObjectIdentifiers.rsaEncryption,
            DERNull.INSTANCE
        )
        val pubKeyX509Bytes = SubjectPublicKeyInfo(algorithmIdentifier, publicKeyBytes).encoded
        val publicKey: JavaSecPublicKey = KeyFactory.getInstance("RSA")
            .generatePublic(X509EncodedKeySpec(pubKeyX509Bytes))
        this.publicKey = publicKey
        this.interfaceId = MessageDigest.getInstance("SHA-256")
            .digest(toPkcs1Bytes())
    }

    /**
     * Creates a [PublicKey] to wrap an RSA [JavaSecPublicKey].
     *
     * @param publicKey The [JavaSecPublicKey] to be wrapped in a [PublicKey].
     */
    constructor(publicKey: JavaSecPublicKey) {
        this.publicKey = publicKey
        this.interfaceId = MessageDigest.getInstance("SHA-256")
            .digest(toPkcs1Bytes())
    }

    //TODO: make these private but make public getters
    private val interfaceId: ByteArray
    private val publicKey: JavaSecPublicKey

    /**
     * Verifies a signature using this [PublicKey].
     *
     * @param signedData The [ByteArray] that has been signed.
     * @param signature The signature of [signedData].
     *
     * @return A [Boolean] indicating whether or not verification was successful.
     */
    fun verifySignature(signedData: ByteArray, signature: ByteArray): Boolean {
        val signatureObj = Signature.getInstance("SHA256withRSA")
        signatureObj.initVerify(this.publicKey)
        signatureObj.update(signedData)
        return signatureObj.verify(signature)
    }

    /**
     * Encrypts the given bytes using this [PublicKey], using OEAP SHA-256 padding.
     *
     * @param clearData The bytes to be encrypted, as a [ByteArray].
     *
     * @return [clearData] encrypted with this [PublicKey].
     */
    fun encrypt(clearData: ByteArray): ByteArray {
        val encryptCipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding")
        val oaepParams = OAEPParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec("SHA-256"), PSource.PSpecified.DEFAULT)
        encryptCipher.init(Cipher.ENCRYPT_MODE, this.publicKey, oaepParams)
        return encryptCipher.doFinal(clearData)
    }

    /**
     * Encodes this [PublicKey] to its PKCS#1-formatted byte representation.
     *
     * @return The PKCS#1 byte representation of this [PublicKey], as a [ByteArray].
     */
    private fun toPkcs1Bytes(): ByteArray {
        val pubKeyX509Bytes = this.publicKey.encoded
        val sPubKeyInfo: SubjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(pubKeyX509Bytes)
        val pubKeyPrimitive: ASN1Primitive = sPubKeyInfo.parsePublicKey()
        return pubKeyPrimitive.encoded // Public key as PKCS#1 bytes
    }
}

File: iosApp/iosApp/key/keys/PublicKey.swift
//
//  PublicKey.swift
//  iosApp
//
//  Created by jimmyt on 11/28/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import CryptoKit
import Foundation

/**
 This is a wrapper around `CryptoKit`'s `SecKey`, with added support for Commuto Interface IDs. The wrapped key is the public key of a 2048-bit RSA private key, and the interface ID of a public key is the SHA-256 hash of its PKCS#1 formatted byte representation.
 
 - Properties:
    - interfaceId: The interface ID of this `PublicKey`, which is the SHA-256 hash of its PKCS#1 byte representation.
    - publicKey: The public RSA key that this `PublicKey` wraps, which is derived from a 2048-bit RSA private key.
 */
struct PublicKey {
    
    let interfaceId: Data
    let publicKey: SecKey
    
    /**
     Creates a `PublicKey` given the PKCS#1-formatted byte representation of a public key.
     
     - Parameter publicKeyBytes: The PKCS#1 bytes of the public key to be created, as `Data`.
     
     - Throws: An `Error` if key creation with the given bytes fails.
     */
    init(publicKeyBytes: Data) throws {
        //Restore public key
        let keyOpts: [String: Any] = [kSecAttrKeyType as String: kSecAttrKeyTypeRSA, kSecAttrKeyClass as String: kSecAttrKeyClassPublic, kSecAttrKeySizeInBits as String: 2048]
        var error: Unmanaged<CFError>?
        guard let publicKey = SecKeyCreateWithData(publicKeyBytes as CFData, keyOpts as CFDictionary, &error) else {
            throw error!.takeRetainedValue() as Error
        }
        self.publicKey = publicKey
        
        //Restore interface id
        let interfaceIdDigest = SHA256.hash(data: publicKeyBytes as NSData)
        var interfaceIdByteArray = [UInt8]()
        for byte: UInt8 in interfaceIdDigest.makeIterator() {
            interfaceIdByteArray.append(byte)
        }
        let interfaceIdBytes = Data(bytes: interfaceIdByteArray, count: interfaceIdByteArray.count)
        self.interfaceId = interfaceIdBytes
    }
    
    /**
     Creates a new `PublicKey` using a public key as a `SecKey`.
     
     - Parameter publicKey: The public `SecKey` that this `PublicKey` wraps.
     
     - Throws: An `Error` if we are unable to compute the inteface ID of `publicKey`.
     */
    init(publicKey: SecKey) throws {
        self.publicKey = publicKey
        var error: Unmanaged<CFError>?
        guard let pubKeyBytes = SecKeyCopyExternalRepresentation(publicKey, &error) else {
            throw error!.takeRetainedValue() as Error
        }
        let interfaceIdDigest = SHA256.hash(data: pubKeyBytes as NSData)
        var interfaceIdByteArray = [UInt8]()
        for byte: UInt8 in interfaceIdDigest.makeIterator() {
            interfaceIdByteArray.append(byte)
        }
        let interfaceIdBytes = Data(bytes: interfaceIdByteArray, count: interfaceIdByteArray.count)
        self.interfaceId = interfaceIdBytes
    }
    
    /**
     Verifies a signature using this `PublicKey`.
     
     - Parameters:
        - signedData: The bytes that have been signed, as `Data`.
        - signature: The signature of `signedData`.
     
     - Returns: A `Bool` indicating whether or not verification was successful.
     
     - Throws: An `Error` if we are unable to complete signature verification.
     */
    func verifySignature(signedData: Data, signature: Data) throws -> Bool {
        let algorithm: SecKeyAlgorithm = .rsaSignatureMessagePKCS1v15SHA256
        var error: Unmanaged<CFError>?
        let success = SecKeyVerifySignature(publicKey, algorithm, signedData as CFData, signature as CFData, &error)
        if !success {
            throw error!.takeRetainedValue() as Error
        }
        return success
    }
    
    /**
     Encrypts the give bytes using this `PublicKey`, using OEAP SHA-256 padding.
     
     - Parameter clearData: The bytes to be encrypted, as `Data`.
     
     - Returns: `clearData` encrypted with this `KeyPair`'s public key.
     
     - Throws: An `Error` if encryption fails.
     */
    func encrypt(clearData: Data) throws -> Data {
        let algorithm: SecKeyAlgorithm = .rsaEncryptionOAEPSHA256
        var error: Unmanaged<CFError>?
        guard let cipherData = SecKeyCreateEncryptedData(publicKey, algorithm, clearData as CFData, &error) as Data? else {
            throw error!.takeRetainedValue() as Error
        }
        return cipherData
    }
    
    /**
     Encodes this `PublicKey` to its PKCS#1-formatted byte representation.
     
     - Returns: The PKCS#1 byte representation of this `PublicKey`, as `Data`.
     
     - Throws: An `Error` of encoding fails.
     */
    func toPkcs1Bytes() throws -> Data {
        var error: Unmanaged<CFError>?
        guard let pubKeyBytes = SecKeyCopyExternalRepresentation(self.publicKey, &error) else {
            throw error!.takeRetainedValue() as Error
        }
        return pubKeyBytes as Data
    }
    
}
